File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 4 - Simple Linear Regression/Simple_Linear_Regression/SimpleLinearRegression.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jan  3 19:24:47 2019

@author: EliteBook
"""

#Simple Linear Regression

# Data Preprocessing Template

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Salary_Data.csv')
X = dataset.iloc[:, :-1].values
Y = dataset.iloc[:, 1].values

# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 1/3, random_state = 0)


#We do not need to apply feature scaling for simple linear regression
# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

#Fitting Simple Linear Regression To Training Set
from sklearn.linear_model import LinearRegression
#creating object of LinearRegression class
regressor = LinearRegression() 
#fitting training set to regressor
regressor.fit(X_train, Y_train)


#predicting test results
Y_Pred = regressor.predict(X_test)
#compare values of Y_pred (model predicted values) and Y_test(actual values)

#Visualising Training Set Results
plt.scatter(X_train, Y_train, color= 'red')
plt.plot(X_train, regressor.predict(X_train), color='blue')
plt.title('Salary vs. Experience (Training Set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.show()

#Visualising Test Set Results
plt.scatter(X_test, Y_test, color= 'red')
plt.plot(X_train, regressor.predict(X_train), color='blue')
plt.title('Salary vs. Experience (Test Set)')
plt.xlabel('Years of Experience')
plt.ylabel('Salary')
plt.show()


#model is making some very good predictions those new employees salaries most of the salaries lie on the best fitting line or are close to the best fitting line

File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 4 - Simple Linear Regression/Simple_Linear_Regression/SimpleLinearRegression.R
#Simple Linear Regression

# Importing the dataset
dataset = read.csv('Salary_Data.csv')

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Salary, SplitRatio = 2/3)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

#Fitting Simple Linear Regression to Training Set
#lm is used to fit linear models
#formula = Salary ~ YearsExperience (Salary is dependent on YearsExperience)
regressor =lm( formula = Salary ~ YearsExperience , data = training_set)

#type summary(regressor) in console to get standard error, r squared value etc after implementing above step
#the P-value is another indicator of the statistical significance because the lower the p value is the more significant your independent variable is going
# to be. That is the more impact the more effect your independent variable is going to have on the dependent variable.
#And usually a good threshold for the P-value is five percent which means that when you are below 5 percent the independent variable is highly significant.
#And when we are over 5 percent that means that it's less significant.

#Predicting the Test Set Results
Y_Pred = predict(regressor, newdata = test_set)

#Visualising the Training Set Results
#install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x= training_set$YearsExperience, y = training_set$Salary), colour = 'red') +
  geom_line(aes( x = training_set$YearsExperience, y = predict( regressor, newdata = training_set)), colour = 'blue') +
  ggtitle('Salary vs Experience (Train Set)') +
  xlab('Years of Experience') +
  ylab('Salary')

#Visualising the Test Set Results
#install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x= test_set$YearsExperience, y = test_set$Salary), colour = 'red') +
  geom_line(aes( x = training_set$YearsExperience, y = predict( regressor, newdata = training_set)), colour = 'blue') +
  ggtitle('Salary vs Experience (Test Set)') +
  xlab('Years of Experience') +
  ylab('Salary')